### PR TITLE
[#3752] CloseWSGIInput middleware + update in ResourceUploader

### PR DESCRIPTION
**Pull request by smotornyuk**
After some investigation of #3752  i found that webob  [creates copy](https://github.com/Pylons/webob/blob/1.0.8/webob/request.py#L546) of request body in some cases(details are not so relevant here). This results in temporary files that still opened until server restart. I suggest additional middleware,that will close such files. 
And, thanks to @curtisyallop for solution, this PR contains fix for ResourceUpload class - it previously relied on automatic file closing as well. Now it performed via context manager

### CODE REVIEW

File: ckan/lib/uploader.py [hunk 1]
```diff
@@ -277,25 +277,25 @@ def upload(self, id, max_size=10):
...
-                    break
-                output_file.write(data)
-                if current_size > max_size:
-                    os.remove(tmp_filepath)
-                    raise logic.ValidationError(
-                        {'upload': ['File upload too large']}
-                    )
+            with open(tmp_filepath, 'wb+') as output_file:
+                with self.upload_file:
+                    self.upload_file.seek(0)
```
Comment (TkTech) [inline ckan/lib/uploader.py]: How is `seek` working on a stream?

Comment (smotornyuk) [inline ckan/lib/uploader.py]: It's not a stream, it is file descriptor

Comment (TkTech) [inline ckan/lib/uploader.py]: Right, of an unbuffered stream?

Comment (smotornyuk) [inline ckan/lib/uploader.py]: But what if `self.upload_file` is a StringIO object(this is case for test and, probably, that could be used for programmatical resource creation)? In that case you need to reset pointer and then copy buffer content into file

**Comment by neilmayhew** [comment]
@TkTech Any chance of getting this finished up and merged soon?

@smotornyuk I think it would be helpful to squash all the commits into one, and rebase the branch onto the current state of `master`.

**Comment by TkTech** [comment]
There are some circle.ci tests failing but I can't see how they were related to the changes in this PR. Will merge tomorrow if there are no further comments.

**Comment by smotornyuk** [comment]
This errors came from current migration to flask(i added fresh changes into this PR).

File: ckan/lib/uploader.py [hunk 2]
```diff
@@ -277,25 +277,25 @@ def upload(self, id, max_size=10):
...
+
+                    if not data:
+                        break
+                    output_file.write(data)
+                    if current_size > max_size:
+                        os.remove(tmp_filepath)
+                        raise logic.ValidationError(
+                            {'upload': ['File upload too large']}
+                        )
+                self.upload_file.close()
```
Comment (neilmayhew) [inline ckan/lib/uploader.py]: Isn't there a possibility of a file leak here, eg if the 'upload too large' exception is thrown? Shouldn't the code be wrapped in a `try ... finally` block?

Comment (curtisyallop) [reply ckan/lib/uploader.py]: Sometimes "upload" is type "file" but for small files, ckan uses type StringO (cStringIO.OutputType) which does not implement __enter__/__exit__ so we cannot use "with:". smotornyuk's previous version used "with:" but we had to switch back to ".close()".

Comment (neilmayhew) [reply ckan/lib/uploader.py]: Yes, I was talking about leaking `self.upload_file`, not the file returned by `open(tmp_filepath)`.

Comment (tino097) [reply ckan/lib/uploader.py]: i see what you meant, i thought we have `file` object for upload too

Comment (smotornyuk) [reply ckan/lib/uploader.py]: Yeah, it has sense. @neilmayhew , thanks, i'll check memory leak here(i believe, this piece will cause it, but checks are always valuable:) ) and then i'll update this piece as you suggested

**Comment by smotornyuk** [comment]
@neilmayhew , you are right, temporary file not removed in case of ValidationError. 

And you may be interested in little change in middleware - now actual file descriptor from request checked rather than `cgi.FileField` itself, because I noticed that it treated like falsy value. Not sure, whether it's because of some update, or there was such behavior initially(in that case, fix shouldn't be working previously :thinking: )

**Comment by curtisyallop** [comment]
@smotornyuk , I can confirm that the falsey behavior was there initially. And the previous fix did not fix the leak. Regular "upload" values of type cgi.FieldStorage evaluate to false for us on ckan 2.3.5 and 2.6 (we run both versions right now). I reported this falsey problem earlier. It used to be here but seems to have disappeared from the githib history. I made a fix and you incorporated this suggestion - to change the code from "if upload:" to "if upload is not None:" (this change also disappeared from github history). Unfortunately my fix, while fixing the falsey problem, had a bug in it - creating url links now failed. At the time, I suggested that the falsey problem might be specific to 2.3.5 but actually it was not.

So to summarize possible values for "upload": "upload" can be None (non-upload regular ckan web site requests), empty string (update resource, create link) or type cgi.FieldStorage (regular uploads). And "if upload:" evaluates to false for type cgi.FieldStorage.

Your latest change looks great. Thanks for your help with this.

**Comment by curtisyallop** [comment]
To test the upload, I propose 5 cases to be tested: (1) Small file (eg 5 bytes), (2) Large file (eg 60MB), (3) Very large file (eg 1GB) (4) Create link (5) Regular site accesses to any page. To test that the leak is fixed, I find the following line works well: `watch -n .2 'df -h; echo; sudo lsof -c python -c httpd -c nginx | grep /tmp'`. lsof shows the non-closed '(deleted)' file handles eating up tmp space. df shows available free disk space dropping. 1GB uploads work well to see space getting eaten up. Note that small files (eg 5 bytes) and large files (eg 60MB) are handled differently in lib/uploader.py.

File: ckan/config/middleware/common_middleware.py [hunk 3]
```diff
@@ -28,6 +29,30 @@ def __call__(self, environ, start_response):
...
+
+    This middleware is supposed to close such files after each request.
+    '''
+    def __init__(self, app, config):
+        self.app = app
+
+    def __call__(self, environ, start_response):
+        wsgi_input = environ['wsgi.input']
+        if isinstance(wsgi_input, FakeCGIBody):
+            upload = wsgi_input.vars.get('upload')
```
Comment (wardi) [inline ckan/config/middleware/common_middleware.py]: These file objects could be passed as any parameter, not just 'upload'. Can you make this fix more general?

**Comment by smotornyuk** [comment]
Updated. 
PS. i checked flask_app - it doesn't has such problem, so middleware should be applied only to pylon

File: ckan/config/middleware/common_middleware.py (hunk 3 above)
Comment (curtisyallop) [reply ckan/config/middleware/common_middleware.py]: Good point. For example the manage-organization image/icon upload is called "image_upload".

File: ckan/lib/uploader.py [hunk 4]
```diff
@@ -22,6 +22,21 @@
 _max_image_size = None
 
 
+def _copy_file(input_file, output_file, max_size):
+    input_file.seek(0)
+    current_size = 0
+    while True:
+        current_size = current_size + 1
+        # MB chunks
+        data = input_file.read(2**20)
```
Comment (wardi) [inline ckan/lib/uploader.py]: since you've moved this code, do you mind giving `2**20` a name? Just that and a rebase or merge to resolve the conflict and I'll merge, Thanks!

**Comment by smotornyuk** [comment]
[Updated](https://github.com/smotornyuk/ckan/blob/e14ede256d93323db2f6cff70f18d7cf183b0a5f/ckan/lib/uploader.py#L18-L33)

**Comment by wardi** [comment]
@smotornyuk thank you!
